### PR TITLE
Health bar

### PR DESCRIPTION
**Pull request by Callmore**
IT'S A HEALTH BAR HOLY SHID

## Media

image
image

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl:
- add: Status UI showing health, role and timer.

### CODE REVIEW

**Comment by Callmore** [comment]
wait hold on i forgor to add it to default hud

**Comment by Callmore** [comment]
also @Simyon264 please actually make sure it works thanks

**Comment by Coolsurf6** [comment]
Wait actually how huge is this, WE NEED MEDIIIAAAA

**Comment by Simyon264** [comment]
image
Testing this locally, the timer seems to not update correctly once the round starts?

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+using Robust.Shared.Timing;
+
+namespace Content.Client._SSS.UserInterface;
+
+public sealed class SSSStatusUIController : UIController, IOnSystemChanged<DamageableSystem>
+{
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
+    [Dependency] private readonly IResourceCache _resourceCache = default!;
+    [Dependency] private readonly IPlayerManager _playerManager = default!;
+    private ISawmill _log = default!;
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:25]: I would prefer this to be spaced out some more. Preferably below the UI system dependencies.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        base.Initialize();
+
+        _log = Logger.GetSawmill("StatusUI");
+
+        SubscribeNetworkEvent<SuspicionRuleTimerUpdate>(UpdateTimerEnd);
+        SubscribeNetworkEvent<SuspicionRulePreroundStarted>(PreroundStarted);
+        SubscribeNetworkEvent<SuspicionRuleUpdateRole>(UpdateRoleDisplay);
+        SubscribeNetworkEvent<SuspicionRulePlayerSpawn>(UpdatePlayerSpawn);
+
+        _log.Info($"{nameof(SSSStatusUIController)} loaded.");
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:44]: Make this a debug log

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    public void UpdateHealth(Entity<DamageableComponent> ent)
+    {
+
+        if (!EntityManager.TryGetComponent<MobStateComponent>(ent, out var mobState))
+            return;
+
+        if (mobState.CurrentState == MobState.Dead)
+        {
+            SetHealthBarCustom("DEAD", 0, 100);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:94]: Localize this string.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+        if (!EntityManager.TryGetComponent<MobThresholdsComponent>(ent, out var mobThresholds))
+            return;
+
+        if (_mobThreshold is null)
+            return;
+
+        var maxHp = _mobThreshold.GetThresholdForState(ent, MobState.Critical, mobThresholds);
+        var hp = Math.Max(Math.Ceiling((maxHp - ent.Comp.TotalDamage).Double()), 0);
+        // var hp = Math.Max((maxHp - ent.Comp.TotalDamage).Float(), 0);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:106]: Commented out?

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        ui.HealthBar.MaxValue = maxValue;
+        ui.HealthBar.Value = value;
+        _queuedHealth = null;
+
+        return true;
+    }
+
+    private void SetHealthBar(float hp, float maxHp)
+    {
+        SetHealthBarCustom($"\u2665 {hp}", hp, maxHp);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:147]: Localize this too.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            ts = TimeSpan.Zero;
+        }
+
+        // nice job c#, TimeSpan.ToString doesn't support having no leading zeros
+        ui.TimerText.Text = $"{ts.Minutes}:{ts.Seconds:00}";
+    }
+
+    public void UpdateTimerEnd(SuspicionRuleTimerUpdate ev, EntitySessionEventArgs args)
+    {
+        _log.Info($"WHAT AT {ev.EndTime}");
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:168]: Make this debug too, preferably also more descriptive of what is being logged?

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        _lastEndTime = ev.EndTime;
+    }
+
+    public void PreroundStarted(SuspicionRulePreroundStarted ev, EntitySessionEventArgs args)
+    {
+        _lastEndTime = ev.PreroundEndTime;
+        SetRoleCustom("Preround", Color.DarkGray);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:175]: Localize this string too.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        ui.RoleBG.PanelOverride = new StyleBoxTexture(_roleStyleBox) { Modulate = color };
+        ui.RoleText.Text = role;
+        _queuedRole = null;
+
+        return true;
+    }
+
+    private void SetRoleToPreround()
+    {
+        SetRoleCustom("Preround", Color.DarkGray);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:212]: Localize this string too.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    private void SetRoleToObserbing()
+    {
+        SetRoleCustom("Obserbing", Color.DarkGray);
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:217]: Localize this string too.

File: Content.Client/_SSS/UserInterface/SSSStatusUIController.cs [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    }
+
+    public void UpdateRoleDisplay(SuspicionRuleUpdateRole ev, EntitySessionEventArgs args)
+    {
+        // ui.RoleBG.PanelOverride = new StyleBoxTexture(_roleStyleBox) { Modulate = Color.FromName(ev.NewRole.GetRoleColor()) };
+        // ui.RoleText.Text = ev.NewRole.ToString();
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/SSSStatusUIController.cs:223]: This is also commented out, just remove if its not needed.

File: Content.Client/_SSS/UserInterface/Widgets/SSSStatusGui.xaml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    xmlns:graphics="clr-namespace:Robust.Client.Graphics;assembly=Robust.Client"
+    VerticalAlignment="Bottom"
+    HorizontalAlignment="Right"
+    >
+
+    <!--
+    VerticalExpand="True"
+    Orientation="Horizontal"
+    -->
+    <Control>
```
Comment (Simyon264) [inline Content.Client/_SSS/UserInterface/Widgets/SSSStatusGui.xaml:14]: Remove commented out code.

File: Content.Server/_SSS/SuspicionGameRule/SuspicionRuleSystem.Spawning.cs [hunk 12]
```diff
@@ -26,6 +27,8 @@ namespace Content.Server._SSS.SuspicionGameRule;
 
 public sealed partial class SuspicionRuleSystem
 {
+    [Dependency] private readonly GameTicker _gameTicker = default!;
```
Comment (Simyon264) [inline Content.Server/_SSS/SuspicionGameRule/SuspicionRuleSystem.Spawning.cs:30]: Put this dependency with the other dependencies in SuspicionRuleSystem.cs

**Review by Simyon264 — CHANGES_REQUESTED** [review]
very sus and very cool

**Comment by Simyon264** [comment]
Lets go try this

**Pull request merged**